### PR TITLE
Add LocaleProvider for Knp and RequestLocaleProvider

### PR DESCRIPTION
**Pull request by franmomu**
## Subject

There are two parts in this PR.

The first commit introduces `LocaleProviderInterface` to be able to fetch the locale from the request or if it's not present will return the default locale.

With this interface we remove the responsibility from `AbstractTranslatableAdminExtension` to the `RequestLocaleProvider` service.

The second commit is adding `LocaleProvider` service when `knplabs` is enabled, this service decorates `LocaleProviderInterface` (from Knplabs) and when the request is from Sonata, it returns the locale from `RequestLocaleProvider` and if is not from Sonata, it uses the decorated service.

This service is necessary to be able to load the proper content translation in sonata using knplabs.

I am targeting this branch, because these changes are BC.

## Changelog

```markdown
### Added
- Added `LocaleProviderInterface` and `RequestLocaleProvider` to get the locale based on the parameter from the URL or the default configured one
- Added `LocaleProvider` service when `knplabs` is enabled to be able to show the content in the proper language
```

### CODE REVIEW

**Comment by franmomu** [comment]
friendly ping @sonata-project/contributors if you can take a look, I think with this one, once is merged in `master` we can release a next major.

**Review by phansys — COMMENTED** [review]
IMO, these changes require an upgrade note.

**Comment by VincentLanglet** [comment]
> friendly ping @sonata-project/contributors if you can take a look, I think with this one, once is merged in `master` we can release a next major.

- Fully typehinted
- Phpstan level max (with strict rules) and psalm 2
- final keyword
- next major resolved
- no more bc-break wanted

Indeed, I feel like this project is ready

**Comment by jordisala1991** [comment]
> > friendly ping @sonata-project/contributors if you can take a look, I think with this one, once is merged in `master` we can release a next major.
> 
> * Fully typehinted
> * Phpstan level max (with strict rules) and psalm 2
> * final keyword
> * next major resolved
> * no more bc-break wanted
> 
> Indeed, I feel like this project is ready

I think we should take a look at deprecations on Symfony 5.4, if we have to break BC to be deprecation free, this is the time.

**Comment by franmomu** [comment]
By the way, I think we can mark the admin extensions as internal since they are not meant to be used individually. I'll do it in `master`.

File: tests/Provider/Knplabs/LocaleProviderTest.php [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+final class LocaleProviderTest extends TestCase
+{
+    /**
+     * @dataProvider provideValues
+     */
+    public function testUsesTheProperProviderBasedOnRequest(
+        string $expectedLocale,
+        string $knpLocale,
+        string $sonataLocale,
+        bool $isSonataEnabled
```
Comment (phansys) [inline tests/Provider/Knplabs/LocaleProviderTest.php]:
```suggestion
        bool $isAdminEnabled
```

File: tests/Provider/Knplabs/LocaleProviderTest.php [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    ): void {
+        $request = new Request();
+
+        if ($isSonataEnabled) {
```
Comment (phansys) [inline tests/Provider/Knplabs/LocaleProviderTest.php]:
```suggestion
        if ($isAdminEnabled) {
```

**Review by phansys — DISMISSED** [review]

**Review by VincentLanglet — APPROVED** [review]

**Review by phansys — APPROVED** [review]

**Pull request merged**